configmap/informer/informed_watcher.go

```diff
@@ -126,31 +130,56 @@ func (i *InformedWatcher) WatchWithDefault(cm corev1.ConfigMap, o ...configmap.O
 	i.Watch(cm.Name, o...)
 }
 
-// Start implements Watcher.
-func (i *InformedWatcher) Start(stopCh <-chan struct{}) error {
-	// Pretend that all the defaulted ConfigMaps were just created. This is done before we start
-	// the informer to ensure that if a defaulted ConfigMap does exist, then the real value is
-	// processed after the default one.
+func (i *InformedWatcher) triggerAddEventForDefaultedConfigMaps(addConfigMapEvent func(obj interface{})) {
 	i.ForEach(func(k string, _ []configmap.Observer) error {
 		if def, ok := i.defaults[k]; ok {
-			i.addConfigMapEvent(def)
+			addConfigMapEvent(def)
 		}
 		return nil
 	})
+}
+
+func (i *InformedWatcher) getConfigMapNames() []string {
+	var configMaps []string
+	i.ForEach(func(k string, _ []configmap.Observer) error {
+		configMaps = append(configMaps, k)
+		return nil
+	})
+	return configMaps
+}
+
+// Start implements Watcher.
+func (i *InformedWatcher) Start(stopCh <-chan struct{}) error {
+	// using the synced callback wrapper around the add event handler will allow the caller
+	// to wait for the add event to be processed for all configmaps
+	s := NewSyncedCallback(i.getConfigMapNames(), i.addConfigMapEvent)
+	addConfigMapEvent := func(obj interface{}) {
+		configMap := obj.(*corev1.ConfigMap)
+		s.Call(obj, configMap.Name)
+	}
+	// Pretend that all the defaulted ConfigMaps were just created. This is done before we start
+	// the informer to ensure that if a defaulted ConfigMap does exist, then the real value is
+	// processed after the default one.
+	i.triggerAddEventForDefaultedConfigMaps(addConfigMapEvent)
 
-	if err := i.registerCallbackAndStartInformer(stopCh); err != nil {
+	if err := i.registerCallbackAndStartInformer(addConfigMapEvent, stopCh); err != nil {
 		return err
 	}
 
-	// Wait until it has been synced (WITHOUT holing the mutex, so callbacks happen)
+	// Wait until the shared informer has been synced (WITHOUT holing the mutex, so callbacks happen)
 	if ok := cache.WaitForCacheSync(stopCh, i.informer.Informer().HasSynced); !ok {
 		return errors.New("error waiting for ConfigMap informer to sync")
 	}
 
-	return i.checkObservedResourcesExist()
+	if err := i.checkObservedResourcesExist(); err != nil {
+		return err
+	}
+
+	// Wait until all config maps have been at least initially processed
+	return s.WaitForAllKeys(stopCh)
```

Comment:
This changes the contract a bit since prior you could run this watcher with supplied defaults and/or missing config maps.

I think being stricter is good but we should document this in the godoc

Reply:
i'm not sure the contract has changed:

* if you supply defaults, it won't wait (as you commented above) (verified via [TestWatchMissingOKWithDefaultOnStart](https://github.com/knative/pkg/blob/8a8de5fc8c72f358ad3e7b5e29f4dfb8a264892d/configmap/informer/informed_watcher_test.go#L300)
* if you don't supply the config maps it expects, it'll fail before this change (checkObservedResourcesExist will return an error i think?), i think [TestWatchMissingFailsOnStart](https://github.com/knative/pkg/blob/8a8de5fc8c72f358ad3e7b5e29f4dfb8a264892d/configmap/informer/informed_watcher_test.go#L229) verifies that

Reply:
I missed the `checkObservedResourcesExist` bit - thanks for pointing that out